### ISSUE TITLE
support to configure other repositories to check for newer versions

### ISSUE THREAD

**Issue by nkiesel** [ben-manes/gradle-versions-plugin#274]
Hi,

our gradle builds are normally only allowed to resolve dependencies against our local artifactory based repo (to make sure developers only use "blessed" jars/versions).  Therefore, I always end up temporarily adding "jcenter()" at to top of the repo list before I use this wonderful plugin.  I wonder if you could add a configuration that would automate that?  Something like
```
dependencyUpdates.additionalRepositories {
    jcenter()
}
```

**Comment by ben-manes** [comment]
Does https://github.com/ben-manes/gradle-versions-plugin/issues/98 help?

**Comment by nkiesel** [comment]
Hmm, looks promising.  However, my understanding is that the order of repos matter in Gradle and given that our local artifactory is basically a heavily trimmed jcenter() I believe that jcenter needs to be added to the top of the repos list.  Looks to me that
```
// add jcenter() to repositories when looking for possible dependency upgrades
tasks.dependencyUpdates.doFirst {
    repositories {
	jcenter()
    }
}
```
instead adds it to the bottom.  Anything I can do about that?

**Comment by ben-manes** [comment]
You might be able to rewrite the order by playing with the collection methods of the api.

https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/dsl/RepositoryHandler.html

**Comment by nkiesel** [comment]
trying, but so far nothing worked.  Which most likely is caused by my ignorance.  I tried `repositories.addFirst(jcenter())` but that says "Could not find method jcenter() for arguments [] on task ':dependencyUpdates' of type com.github.benmanes.gradle.versions.updates.DependencyUpdatesTask."
And `repositories { addFirst(jcenter()) }` results in "Cannot add a repository with name 'BintrayJCenter' as a repository with that name already exists."

**Comment by ben-manes** [comment]
The convenience methods add themselves. Maybe instead copy the list, clear, and add in order. Or sort it by a comparator. Or if you only need certain ones, remove. See 
https://github.com/ben-manes/gradle-versions-plugin/issues/116 as an example.

**Comment by nkiesel** [comment]
Ah, that makes sense. Let me try tomorrow, I will update this issue after further tests.

**Comment by nkiesel** [comment]
I now have the following:
```
// add jcenter() as first element to repositories when looking for possible
// dependency upgrades
tasks.dependencyUpdates.doFirst {
    // first add jcenter to the end of the list
    repositories {
	jcenter()
    }
    // get the last element which is the just added jcenter
    def jcenter = repositories[-1]
    // remove jcenter. The repo container guarantees that names are unique, so
    // this will always work correctly
    repositories.removeIf { it.name == jcenter.name }
    // and now add again as first repository
    repositories.addFirst(jcenter)
}
```
The good news is that this technically seems to work (I listed the repos and jcenter indeed ends up as the first). The bad news is that the result of running with this configuration is very different than with the configuration where I add jcenter directly into my repo list (and comment out the above).  The former misses available updated versions that the latter finds.
One aspect that might be relevant here is that we have a multi-module setup and that the standard repositories block is thus inside `allProjects`, while the code block above is outside.

**Comment by nkiesel** [comment]
Hmm. I (by accident) left `jcenter()` in my standard `repositories` _and_ had that code above included, and that did not produce a "cannot add jcenter; duplicate name".  So I printed the repo list both inside the standard `repositories` closure and inside `tasks.dependencyUpdates.doFirst` and I see the output from the standard one running multiple times (presumably for every submodule) and then the output of  `tasks.dependencyUpdates.doFirst`.  This makes me think that the list was already cloned for all the modules and the change from  `tasks.dependencyUpdates.doFirst` thus is effectively ignored.

**Comment by ben-manes** [comment]
You would want to put it in `allProjects` since this plugin uses your project repositories. It does get quirky hacking the repository listing which is why we stopped when we encountered odd bugs. So it is delicate.

Are you sure that adding `jcenter` at the end is not good enough? The resolution should query every repository when looking for the latest.

**Comment by nkiesel** [comment]
I think I figured it out: wrapping the repo modification inside a `allprojects` does the trick.  For the record, my code now looks like this:
```
// add jcenter() as first element to repositories when looking for possible
// dependency upgrades
tasks.dependencyUpdates.doFirst {
    // we need to do this for all projects
    allprojects {
	// first add jcenter to the end of the list
	repositories {
	    jcenter()
	}
	// get the last element which is the just added jcenter
	def jcenter = repositories[-1]
	// remove jcenter. The repo container guarantees that names are unique, so
	// this will always work correctly
	repositories.removeIf { it.name == jcenter.name }
	// and now add again as first repository
	repositories.addFirst(jcenter)
    }
}
```

**Comment by nkiesel** [comment]
Sorry, did not see your `allprojects` before. Good, so that confirms I was right.

And yes: adding it to the beginning seems critical because we have a subset of e.g. commons-lang3 in our local repo, and gradle docu contains the following note:

>  The order of declaration determines how Gradle will check for dependencies at runtime. If Gradle finds a module descriptor in a particular repository, it will attempt to download all of the artifacts for that module from the same repository. You can learn more about the inner workings of Gradle’s resolution mechanism.

Thus, it will never look at jcenter if it is behind our local repo

**Comment by ben-manes** [comment]
True, but because we use a dynamic version I think Gradle would have to query every repository to find out the latest. So `a:b:+` might have 1.0 on the first, and 2.2 on the second. Gradle would have to query to figure that out. It would prefer the first repository all things being equal for explicit versions.

**Comment by nkiesel** [comment]
From gradle docu (emphasis mine):
> If the dependency is declared as a dynamic version (like 1.+), Gradle will resolve this to the highest available concrete version (like 1.2) **in the repository**.

I also see that behavior between `mavenCentral()` and `jcenter()`: mavenCentral has `2.3` as latest version of 'net.sf.opencsv:opencsv' while jcenter has `2.4`. And switching the order of the 2 repos will report either `2.3` or 2.4` as the latest version.

This seems to be a real theoretical problem for this plugin.  It just happens to be that jcenter and mavencentrla are in sync to a very high degree that it might practically not be relevant.

**Comment by ben-manes** [comment]
JCenter is actually pretty bad at being in sync. It won't pull dynamic versions from Maven Central, only proxy point versions. So if a JCenter repository has version 1.x and the user doesn't include Central, they won't see 2.x. But, at least in the past, if they added Central in their list it would query all of them and get the latest.

Gradle keeps changing their repository / dependency resolution rules, so you're probably right. It used to work as I described. They do have to evolve and move fast, so we'll see how things work out.
